crates/rpc/rpc-types/src/eth/block.rs

```diff
@@ -33,6 +45,80 @@ pub struct Block {
     pub base_fee_per_gas: Option<U256>,
 }
 
+impl Block {
+    /// Create a new block response from a [primitive block](reth_primitives::Block), using the
+    /// total difficulty to populate its field in the rpc response.
+    pub fn from_block_full(
+        block: PrimitiveBlock,
+        total_difficulty: U256,
+    ) -> Result<Self, BlockError> {
+        let block_hash = block.header.hash_slow();
+        let header_length = block.header.length();
+        let block_length = block.length();
+        let uncles = block.ommers.into_iter().map(|h| h.hash_slow()).collect();
+
+        let RethHeader {
+            parent_hash,
+            ommers_hash,
+            beneficiary,
+            state_root,
+            transactions_root,
+            receipts_root,
+            logs_bloom,
+            difficulty,
+            number,
+            gas_limit,
+            gas_used,
+            timestamp,
+            mix_hash,
+            nonce,
+            base_fee_per_gas,
+            extra_data,
+        } = block.header;
+
+        let header = Header {
+            hash: Some(block_hash),
+            parent_hash,
+            uncles_hash: ommers_hash,
+            author: beneficiary,
+            miner: beneficiary,
+            state_root,
+            transactions_root,
+            receipts_root,
+            number: Some(U256::from(number)),
+            gas_used: U256::from(gas_used),
+            gas_limit: U256::from(gas_limit),
+            extra_data,
+            logs_bloom,
+            timestamp: U256::from(timestamp),
+            difficulty,
+            mix_hash,
+            nonce: Some(nonce.to_be_bytes().into()),
+            size: Some(U256::from(header_length)),
+        };
+
+        let mut transactions = Vec::with_capacity(block.body.len());
+        for (idx, tx) in block.body.iter().enumerate() {
+            let signed_tx = tx.clone().into_ecrecovered().ok_or(BlockError::InvalidSignature)?;
```

Comment:
This will be slow if `from_block_full` is called many times, I would query the senders table  and pass the senders as an argument to this fn. Else do over par_iter.

Reply:
Good point, having senders as an argument makes sense for this. I'm not sure which blocks would contain transactions where we haven't validated the signatures or performed sender recovery, but we shouldn't return blocks over RPC that we haven't recovered the senders for already